### PR TITLE
Load JRA55 data correctly when time indices wrap around

### LINKED ISSUE #285: BoundsError in `update_field_time_series!`
This might be an Oceananigans problem but a simulation I'm running recently threw a bounds error in update state:

```julia
[ Info: Time: 360 days, n: 51840, Δt: 10 minutes, max|u|: (2.32e+00, 3.14e+00, 1.26e-02) m s⁻¹, extrema(T): (42.83, -1.87) ᵒC, wall time: 71.343 ms
ERROR: LoadError: BoundsError: attempt to access 640×320×2 Array{Float32, 3} at index [1:640, 1:320, 1:39]
Stacktrace:
  [1] throw_boundserror(A::Array{Float32, 3}, I::Tuple{UnitRange{Int64}, UnitRange{Int64}, UnitRange{Int64}})
    @ Base ./abstractarray.jl:737
  [2] checkbounds
    @ ./abstractarray.jl:702 [inlined]
  [3] view
    @ ./subarray.jl:184 [inlined]
  [4] maybeshrink(temparray::Array{…}, a::NCDatasets.Variable{…}, indices::Tuple{…})
    @ DiskArrays /orcd/data/raffaele/001/glwagner/.julia/packages/DiskArrays/ny95C/src/diskarray.jl:290
  [5] getindex_disk_batch!(out::Nothing, a::NCDatasets.Variable{Float32, 3, NCDatasets.NCDataset{…}}, i::Tuple{UnitRange{…}, UnitRange{…}, Vector{…}})
    @ DiskArrays /orcd/data/raffaele/001/glwagner/.julia/packages/DiskArrays/ny95C/src/diskarray.jl:231
  [6] getindex_disk!
    @ /orcd/data/raffaele/001/glwagner/.julia/packages/DiskArrays/ny95C/src/diskarray.jl:258 [inlined]
  [7] getindex_disk
    @ /orcd/data/raffaele/001/glwagner/.julia/packages/DiskArrays/ny95C/src/diskarray.jl:218 [inlined]
  [8] getindex
    @ /orcd/data/raffaele/001/glwagner/.julia/packages/DiskArrays/ny95C/src/diskarray.jl:370 [inlined]
  [9] getindex(::CommonDataModel.CFVariable{…}, ::UnitRange{…}, ::UnitRange{…}, ::Vector{…})
    @ CommonDataModel /orcd/data/raffaele/001/glwagner/.julia/packages/CommonDataModel/vTYnt/src/cfvariable.jl:451
 [10] set!(fts::FieldTimeSeries{…}, path::String, name::String)
    @ ClimaOcean.DataWrangling.JRA55 /orcd/data/raffaele/001/glwagner/ClimaOcean.jl/src/DataWrangling/JRA55.jl:255
 [11] set! (repeats 2 times)
    @ /orcd/data/raffaele/001/glwagner/.julia/packages/ClimaOcean/vd1yD/src/DataWrangling/JRA55.jl:237 [inlined]
 [12] update_field_time_series!(fts::FieldTimeSeries{…}, n₁::Int64, n₂::Int64)
    @ Oceananigans.OutputReaders /orcd/data/raffaele/001/glwagner/.julia/packages/Oceananigans/ohMpu/src/OutputReaders/field_time_series_indexing.jl:269
 [13] update_field_time_series!(fts::FieldTimeSeries{…}, time_index::Time{…})
    @ Oceananigans.OutputReaders /orcd/data/raffaele/001/glwagner/.julia/packages/Oceananigans/ohMpu/src/OutputReaders/field_time_series_indexing.jl:257
 [14] update_model_field_time_series!(atmos::ClimaOcean.OceanSeaIceModels.PrescribedAtmospheres.PrescribedAtmosphere{…}, time::Time{…})
    @ ClimaOcean.OceanSeaIceModels.PrescribedAtmospheres /orcd/data/raffaele/001/glwagner/ClimaOcean.jl/src/OceanSeaIceModels/PrescribedAtmospheres.jl:364
 [15] #update_state!#3
    @ /orcd/data/raffaele/001/glwagner/.julia/packages/ClimaOcean/vd1yD/src/OceanSeaIceModels/time_step_ocean_sea_ice_model.jl:47 [inlined]
 [16] update_state!
    @ /orcd/data/raffaele/001/glwagner/.julia/packages/ClimaOcean/vd1yD/src/OceanSeaIceModels/time_step_ocean_sea_ice_model.jl:45 [inlined]
 [17] time_step!(coupled_model::ClimaOcean.OceanSeaIceModels.OceanSeaIceModel{…}, Δt::Float64; callbacks::Tuple{}, compute_tendencies::Bool)
    @ ClimaOcean.OceanSeaIceModels /orcd/data/raffaele/001/glwagner/ClimaOcean.jl/src/OceanSeaIceModels/time_step_ocean_sea_ice_model.jl:40
 [18] time_step!(sim::Simulation{ClimaOcean.OceanSeaIceModels.OceanSeaIceModel{…}, Float64, Float64, OrderedDict{…}, OrderedDict{…}, OrderedDict{…}})
    @ Oceananigans.Simulations /orcd/data/raffaele/001/glwagner/.julia/packages/Oceananigans/ohMpu/src/Simulations/run.jl:134
 [19] run!(sim::Simulation{ClimaOcean.OceanSeaIceModels.OceanSeaIceModel{…}, Float64, Float64, OrderedDict{…}, OrderedDict{…}, OrderedDict{…}}; pickup::Bool)
    @ Oceananigans.Simulations /orcd/data/raffaele/001/glwagner/.julia/packages/Oceananigans/ohMpu/src/Simulations/run.jl:97
 [20] run!
    @ /orcd/data/raffaele/001/glwagner/.julia/packages/Oceananigans/ohMpu/src/Simulations/run.jl:85 [inlined]
 [21] forward_run!(simulation::Simulation{…}, parameters::Vector{…}, ensemble_index::Int64, iteration::Int64)
    @ Main /orcd/data/raffaele/001/glwagner/ClimaOceanCalibration.jl/examples/north_atlantic_forward_run.jl:44
 [22] top-level scope
    @ /orcd/data/raffaele/001/glwagner/ClimaOceanCalibration.jl/examples/north_atlantic_forward_run.jl:66
 [23] include(fname::String)
    @ Base.MainInclude ./client.jl:489
 [24] top-level scope
    @ REPL[20]:1
in expression starting at /orcd/data/raffaele/001/glwagner/ClimaOceanCalibration.jl/examples/north_atlantic_forward_run.jl:66
Some type information was truncated. Use `show(err)` to see complete types.
```

### PR DESCRIPTION
**Pull request by glwagner**
Still need to add a test.

Closes #285

### CODE REVIEW

File: src/DataWrangling/JRA55.jl [hunk 1]
```diff
@@ -252,7 +252,15 @@ function set!(fts::JRA55NetCDFFTS, path::String=fts.path, name::String=fts.name)
...
     ti = collect(ti)
-    data = ds[name][i₁:i₂, j₁:j₂, ti]
+
+    if issorted(ti)
+        data = ds[name][i₁:i₂, j₁:j₂, ti]
+    else # ti must wrap around 1
+        fwd = sortperm(ti)
+        data = ds[name][i₁:i₂, j₁:j₂, ti[fwd]]
+        data = data[:, :, invperm(fwd)]
+    end
```
Comment (navidcy) [inline src/DataWrangling/JRA55.jl:262]: nice

File: src/DataWrangling/JRA55.jl [hunk 2]
```diff
@@ -252,7 +252,15 @@ function set!(fts::JRA55NetCDFFTS, path::String=fts.path, name::String=fts.name)
 
     ti = time_indices(fts)
     ti = collect(ti)
-    data = ds[name][i₁:i₂, j₁:j₂, ti]
+
+    if issorted(ti)
+        data = ds[name][i₁:i₂, j₁:j₂, ti]
+    else # ti must wrap around 1
```
Comment (navidcy) [inline src/DataWrangling/JRA55.jl:258]: I don't understand what "wrap around 1" mean
I think I understand what you mean when I see the linked issue; but if the issue wasn't there I wouldn't know..
I presume this refers to using forcing from 1 year to force for longer than a year? shall we make the comment more informative?

**Comment by navidcy** [comment]
Given https://github.com/CliMA/ClimaOcean.jl/issues/285#issuecomment-2520984544 let's add a compat entry to NCDatasets in this PR?

**Review by navidcy — APPROVED** [review]
made a few comments

**Comment by navidcy** [comment]
I see there is a test but the test doesn't cover the case where `ti = [3, 4, 5, 1]`, right?

File: src/DataWrangling/JRA55.jl (hunk 2 above)
Comment (glwagner) [reply src/DataWrangling/JRA55.jl:258]: what do you suggest?

**Comment by glwagner** [comment]
> Given [#285 (comment)](https://github.com/CliMA/ClimaOcean.jl/issues/285#issuecomment-2520984544) let's add a compat entry to NCDatasets in this PR?

I think this PR works regardless of the NCDatasets version. However, we only discovered the error with the latest version. But that doesn't matter when we implement the correct algorithm. Unless I am missing something?

**Comment by glwagner** [comment]
> I see there is a test but the test doesn't cover the case where `ti = [3, 4, 5, 1]`, right?

That's the case the test covers; the indices are [Nt-2, Nt-1, Nt, 1]. Nt is much larger than 5 because we load time_indices=Colon() but the tested functionality is the same?

**Comment by navidcy** [comment]
> > I see there is a test but the test doesn't cover the case where `ti = [3, 4, 5, 1]`, right?
> 
> That's the case the test covers; the indices are [Nt-2, Nt-1, Nt, 1]. Nt is much larger than 5 because we load time_indices=Colon() but the tested functionality is the same?

oh I missed that line in the tests -- sorry, all good ignore my comment!

File: src/DataWrangling/JRA55.jl [hunk 3]
```diff
@@ -255,10 +255,14 @@ function set!(fts::JRA55NetCDFFTS, path::String=fts.path, name::String=fts.name)
 
     if issorted(ti)
         data = ds[name][i₁:i₂, j₁:j₂, ti]
-    else # ti must wrap around 1
+    else
+        # The time indices may be cycling past 1; eg ti = [6, 7, 8, 1].
+        # However, DiskArrays does not seem to support loading data with unsorted
+        # indices. So to handle this, we first load the data with sorted time-indices,
+        # and then invert the permutation to obtain the desired cycling time-series.
```
Comment (navidcy) [inline src/DataWrangling/JRA55.jl:262]: excellent

File: src/DataWrangling/JRA55.jl (hunk 2 above)
Comment (navidcy) [reply src/DataWrangling/JRA55.jl]: [b7c6dcf](https://github.com/CliMA/ClimaOcean.jl/pull/287/commits/b7c6dcf2b04cbe3438700cc4661e0f2a3d145d51) is great!

**Comment by glwagner** [comment]
Hmm yea so tests passed this time. I think the failure was just some intermittent server outage or something

**Comment by glwagner** [comment]
I didn't mean to merge this, I meant to merge another PR. Whoops..

**Pull request merged**